nekoyume/node.py

```diff
@@ -0,0 +1,134 @@
+            node = Node(url=url, last_connected_at=datetime.datetime.utcnow())
+            if session:
+                session.add(node)
+                try:
+                    session.commit()
+                except IntegrityError:
+                    node = get_node()
+                    if node is None:
+                        raise
+                    return node
+            return node
+        else:
+            return None
+
+    @classmethod
+    def update(cls, node: 'Node'):
+        """
+        Update recent node list by scrapping other nodes' information.
+        """
+        try:
+            response = get(f"{node.url}{Node.get_nodes_endpoint}")
+        except (ConnectionError, Timeout):
+            return
+        for url in response.json()['nodes']:
+            try:
+                Node.get(url)
+            except (ConnectionError, Timeout):
+                continue
+        db.session.commit()
+
+    def ping(self):
+        try:
+            result = get(f'{self.url}/ping').text == 'pong'
+            if result:
+                self.last_connected_at = datetime.datetime.utcnow()
+            return result
+        except (ConnectionError, Timeout):
+            return False
+
+    @classmethod
+    def broadcast(cls,
+                  endpoint: str,
+                  serialized_obj: typing.Mapping[str, object],
+                  sent_node: typing.Optional['Node']=None,
+                  my_node: typing.Optional['Node']=None,
+                  session: Session=db.session) -> bool:
+        """
+        It broadcast `serialized_obj` to every nodes you know.
+
+        :param        endpoint: endpoint of node to broadcast
+        :param  serialized_obj: object that will be broadcasted.
+        :param       sent_node: sent :class:`nekoyume.models.Node`.
+                                this node ignore sent node.
+        :param         my_node: my :class:`nekoyume.models.Node`.
+                                received node ignore my node when they
+                                broadcast received object.
+        """
+        from .models import Block
+        for node in session.query(cls):
+            if sent_node and sent_node.url == node.url:
+                continue
+            try:
+                if my_node:
+                    serialized_obj['sent_node'] = my_node.url
+                res = post(node.url + endpoint, json=serialized_obj,
+                           timeout=3)
+                if res.status_code == 403:
+                    result = res.json()
+                    try:
+                        block_id = result['block_id']
+                    except KeyError:
+                        continue
+                    query = session.query(Block).filter(
+                        Block.id.between(block_id, serialized_obj['id'])
+                    ).order_by(Block.id)
+                    count = query.count()
+                    while count > 0:
```

Comment:
You don't have to `count(query)` before fetching it, because for the most part we need to fetch it anyway.  Instead, you can loop until a fetched result is empty, e.g.:

```python
while True:
    sync_blocks = query.limit(DEFAULT_BROADCAST_LIMIT).all()
    if not sync_blocks:
        break
```

Reply:
I'd misunderstood the code. Never mind.